services/catalog/src/simcore_service_catalog/db/repositories/services.py

```diff
@@ -205,12 +207,133 @@ async def get_service(
+            ):
+                msg = f"{access_rights} does not correspond to service {new_service.key}:{new_service.version}"
+                raise ValueError(msg)
+
+        async with self.db_engine.begin() as conn:
+            # NOTE: this ensure proper rollback in case of issue
+            result = await conn.execute(
+                # pylint: disable=no-value-for-parameter
+                services_meta_data.insert()
+                .values(**new_service.dict(by_alias=True))
+                .returning(literal_column("*"))
+            )
+            row = result.first()
+            assert row  # nosec
+            created_service = cast(
+                ServiceMetaDataAtDB, ServiceMetaDataAtDB.from_orm(row)
+            )
+
+            for access_rights in new_service_access_rights:
+                insert_stmt = pg_insert(services_access_rights).values(
+                    **access_rights.dict(by_alias=True)
+                )
+                await conn.execute(insert_stmt)
+
+        return created_service
+
+    async def update_service(
+        self, patched_service: ServiceMetaDataAtDB
+    ) -> ServiceMetaDataAtDB:
+
+        stmt_update = (
+            services_meta_data.update()
+            .where(
+                (services_meta_data.c.key == patched_service.key)
+                & (services_meta_data.c.version == patched_service.version)
+            )
+            .values(
+                **patched_service.dict(
+                    by_alias=True,
+                    exclude_unset=True,
+                    exclude={"key", "version"},
+                )
+            )
+            .returning(literal_column("*"))
+        )
+
+        async with self.db_engine.begin() as conn:
+            result = await conn.execute(stmt_update)
+            row = result.first()
+            assert row  # nosec
+        return cast(ServiceMetaDataAtDB, ServiceMetaDataAtDB.from_orm(row))
+
+    # NEW CRUD on services ------
+
+    async def get_service_with_history(
+        self,
+        # access-rights
+        product_name: ProductName,
+        user_id: UserID,
+        # get args
+        key: ServiceKey,
+        version: ServiceVersion,
+    ) -> ServiceWithHistoryFromDB | None:
+
+        stmt_get = get_service_stmt2(
+            product_name=product_name,
+            user_id=user_id,
+            access_rights=AccessRightsClauses.can_read,
+            service_key=key,
+            service_version=version,
+        )
+
+        async with self.db_engine.begin() as conn:
+            result = await conn.execute(stmt_get)
+            row = result.one_or_none()
+
+        if row:
+            stmt_history = get_service_history_stmt(
+                product_name=product_name,
+                user_id=user_id,
+                access_rights=AccessRightsClauses.can_read,
+                service_key=key,
+            )
+            async with self.db_engine.begin() as conn:
```

Comment:
Q: Just curious is there a special reason why you create a new connection and do not reuse the previous one?

Reply:
I do not want to hold the connection unecessarily. Does it make sense?